src/dist_kmeans_init.h

```diff
@@ -198,6 +198,122 @@ namespace dist_custom {
+            auto step2In = step14Out;
+
+            // default value of nRounds used by all steps
+            const size_t nRounds = daal::algorithms::kmeans::init::Parameter(algo._nClusters).nRounds;
+
+            // vector with results of step2 for input into step5
+            std::vector<daal::data_management::NumericTablePtr> s2InForStep5;
+            if(rank == 0) {
+                s2InForStep5.push_back(step2In);
+            }
+
+            // Here we will store the output of step3 for step5
+            daal::services::interface1::SharedPtr<daal::data_management::interface1::SerializationIface> outputOfStep3ForStep5;
+
+            for(size_t iRound = 0; iRound < nRounds; ++iRound) {
+                // run step2 on each rank
+                auto s2res = algo.run_step2Local(input, localNodeData, step2In, false);
+                if(iRound==0) localNodeData = s2res->get(daal::algorithms::kmeans::init::internalResult);
+                auto s2Out = s2res->get(daal::algorithms::kmeans::init::outputOfStep2ForStep3);
+                //  and gather result on root
+                auto s3In = tcvr->gather(s2Out);
+
+                const int S34TAG = 3003;
+                // The input for s4 will be stored in s4In
+                daal::data_management::NumericTablePtr s4In;
+                // run step3 on root and send results to non-roots
+                if(rank == 0) {
+                    auto step3Output = algo.run_step3Master(s3In);
+                    // output of step3 in the last iteration will be used in step5
+                    if (iRound == nRounds - 1) {
+                        outputOfStep3ForStep5 = step3Output->get(daal::algorithms::kmeans::init::outputOfStep3ForStep5);
+                    }
+                    s4In = step3Output->get(daal::algorithms::kmeans::init::outputOfStep3ForStep4, 0);
+                    for(int i=1; i<nRanks; i++) {
+                        tcvr->send(step3Output->get(daal::algorithms::kmeans::init::outputOfStep3ForStep4, i), i, S34TAG); // it can be NULL
+                    }
+                } else { // non-roots get messages with output from step3
+                    s4In = tcvr->recv<daal::data_management::NumericTablePtr>(0, S34TAG);
+                }
+                // if we have a data for step4 then run it
+                if (s4In) {
+                    step14Out = algo.run_step4Local(input, localNodeData, s4In);
+                } else {
+                    step14Out = daal::data_management::NumericTablePtr();
+                }
+
+                // we need to gather all exist results on root, merge them into one table and then share it with all non-roots
+                auto step14OutMaster = tcvr->gather(step14Out);
+                daal::data_management::RowMergedNumericTablePtr step4OutMerged(new daal::data_management::RowMergedNumericTable());
+                if(rank == 0)
+                {
+                    for (int i = 0; i < step14OutMaster.size(); i++)
```

Comment:
See other comment, gather should return result of rank/process i at position i. This code must handle possible NULL pointers.